pkg/registry/softwarecomposition/networkneighbors/strategy.go

```diff
@@ -0,0 +1,121 @@
+	if !ok {
+		return nil, nil, fmt.Errorf("given object is not a Flunder")
+	}
+	return labels.Set(apiserver.ObjectMeta.Labels), SelectableFields(apiserver), nil
+}
+
+// MatchApplicationProfileSummary is the filter used by the generic etcd backend to watch events
+// from etcd to clients of the apiserver only interested in specific labels/fields.
+func MatchNetworkNeighbor(label labels.Selector, field fields.Selector) storage.SelectionPredicate {
+	return storage.SelectionPredicate{
+		Label:    label,
+		Field:    field,
+		GetAttrs: GetAttrs,
+	}
+}
+
+// SelectableFields returns a field set that represents the object.
+func SelectableFields(obj *softwarecomposition.NetworkNeighbors) fields.Set {
+	return generic.ObjectMetaFieldsSet(&obj.ObjectMeta, true)
+}
+
+type networkNeighborsStrategy struct {
+	runtime.ObjectTyper
+	names.NameGenerator
+}
+
+func (networkNeighborsStrategy) NamespaceScoped() bool {
+	return true
+}
+
+func (networkNeighborsStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+}
+
+func (networkNeighborsStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+}
+
+func (networkNeighborsStrategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
+	networkNeighbors := obj.(*softwarecomposition.NetworkNeighbors)
+
+	return validatePorts(networkNeighbors)
+}
+
+// WarningsOnCreate returns warnings for the creation of the given object.
+func (networkNeighborsStrategy) WarningsOnCreate(ctx context.Context, obj runtime.Object) []string {
+	return nil
+}
+
+func (networkNeighborsStrategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+func (networkNeighborsStrategy) AllowUnconditionalUpdate() bool {
+	return false
+}
+
+func (networkNeighborsStrategy) Canonicalize(obj runtime.Object) {
+}
+
+func (networkNeighborsStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
+	networkNeighbors := obj.(*softwarecomposition.NetworkNeighbors)
+
+	return validatePorts(networkNeighbors)
+}
+
+// WarningsOnUpdate returns warnings for the given update.
+func (networkNeighborsStrategy) WarningsOnUpdate(ctx context.Context, obj, old runtime.Object) []string {
+	return nil
+}
+
+func validatePorts(networkNeighbors *softwarecomposition.NetworkNeighbors) field.ErrorList {
+	for _, ingress := range networkNeighbors.Spec.Ingress {
+		for _, networkPort := range ingress.Ports {
+			if *networkPort.Port < 0 || *networkPort.Port > 65535 {
+				return field.ErrorList{field.Invalid(field.NewPath("spec").Child("ingress").Child("ports"), *networkPort.Port, "port must be in range 0-65535")}
+			}
+
+			expectedPortName := fmt.Sprintf("%s-%d", networkPort.Protocol, *networkPort.Port)
+			if networkPort.Name != expectedPortName {
+				return field.ErrorList{field.Invalid(field.NewPath("spec").Child("ingress").Child("ports"), *networkPort.Port, fmt.Sprintf("port name must be in format {protocol}-{port}, expected name: %s", expectedPortName))}
+			}
```

Comment:
issue: these two logical validations are not separated in code

These should be separate validators. Something like `validatePortNumber` and `validatePortName`. Take a look how the original validators are structured in `pkg/apis/softwarecomposition/validation/validation.go`. They traverse the objects from the `Object` to `ObjectSpec`. Whenever you go down a level to a new object type, it should be a new validator. In this case:
- `ValidateNetworkNeighbors(nn *softwarecomposition.NetworkNeighbors, fldPath *field.Path)`
- `ValidateNetworkNeighborsSpec(nns *softwarecomposition.NetworkNeighborsSpec, fldPath *field.Path)`
- `ValidateNetworkNeighbor(nn *softwarecomposition.NetworkNeighbor, fldPath *field.Path)`
- `ValidatePort(p *softwarecomposition.Port, fldPath *field.Path)`
etc.

Since these validators would take a field path as an argument, it should also eliminate the problem of duplicating the same code when you validate the `NetworkEntry` / `NetworkNeighbor`. Since you could just pass a `fieldPath.Child("ingress")` or `fieldPath.Child("egress")` depending on what you are passing for validation.